### PR TITLE
osd: Read balancer for devices with different sizes (performance)

### PR DESCRIPTION
**Pull request by JoshSalomon**
Read Balancer: add support for better balancing for devices with different sizes.

When there are devices with different sizes, and the capacity is balanced (very important from an operational perspective) larger devices take more load than smaller devices and as a result immediately create the weakest link in the chain effect and reduce the maximal cluster bandwidth. This PR aims to add more load on the smaller devices (by adding primaries) and less load on the larger devices (by reducing the number of primaries) and as a result create more balanced load on the OSDs and improve the maximal cluster bandwidth.
For the calculations to work, we need to have a good estimation of the IOs read/write ratio per pool, therefore this PR adds a new pool parameter `read_ratio`
Under the assumption that devices with the same technology have the same bandwidth irrelevant to their size, this PR improves the overall cluster performance in many cases. 

This feature is opt-in and can be activated now only via osdmaptool 

Documentation will be provided in a different PR

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
@markhpc this PR is intended as a performance improvement to the read balancer, and it will be an optional configuration in the Squid offline tool (osdmaptool).

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -5196,6 +5273,229 @@ int OSDMap::calc_desired_primary_distribution(
...
+    int64_t read_ratio = 0;
+    {
+      // new scope since it is not allowed to use def_read_ratio after std::move
+      uint64_t def_read_ratio = cct->_conf.get_val<uint64_t>("osd_pool_default_read_ratio");
+      read_ratio = pool->opts.value_or<int64_t>(pool_opts_t::key_t::READ_RATIO, std::move(def_read_ratio));
+    }
+    int write_ratio = 100 - read_ratio;
+    ldout(cct, 30) << __func__ << " Pool: " << pid << " read ratio: " << read_ratio << " write ratio: " << write_ratio << dendl;
+    int num_osds = osds.size();
+    if (pg_num != (pool->get_pgp_num_mask() + 1)) {
```
Comment (ljflores) [inline src/osd/OSDMap.cc:5335]: Seeing this compile warning. One of them should be cast as (int) or (uint):
```
/home/lflores/ceph/src/osd/OSDMap.cc:5335:16: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
 5335 |     if (pg_num != (pool->get_pgp_num_mask() + 1)) {
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -6226,39 +6517,181 @@ int OSDMap::calc_read_balance_score(CephContext *cct, int64_t pool_id,
...
+  }
+
+  map<uint64_t,set<pg_t>> pgs_by_osd;
+  map<uint64_t,set<pg_t>> prim_pgs_by_osd;
+  map<uint64_t,set<pg_t>> acting_prims_by_osd;
+
+  pgs_by_osd = tmp_osd_map.get_pgs_by_osd(cct, pool_id, &prim_pgs_by_osd, &acting_prims_by_osd);
+  auto num_osds = pgs_by_osd.size();
+  int64_t num_pg_osd_legs = 0;
+  for (int i = 0 ; i < num_osds ; i++) {
```
Comment (ljflores) [inline src/osd/OSDMap.cc:6579]: Also seeing a compile warning on this line. `num_osds` should probably be cast to an `int`:
```
/home/lflores/ceph/src/osd/OSDMap.cc:6579:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 6579 |   for (int i = 0 ; i < num_osds ; i++) {
      |                    ~~^~~~~~~~~~
```

**Comment by ljflores** [comment]
Testing the PR here:
https://pulpito.ceph.com/lflores-2024-01-29_16:07:33-rados-wip-lflores-testing-2024-01-29-0542-distro-default-smithi/

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test make check

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (JoshSalomon) [reply src/osd/OSDMap.cc:5335]: Changed declaration to `auto`

**Comment by ljflores** [comment]
Tested here: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcl9Ab9HaE1942-wip-lflores-testing-2024-01-29-0542

**Review by ljflores — APPROVED** [review]

**Comment by ljflores** [comment]
jenkins test windows

File: src/osd/OSDMap.cc [hunk 3]
```diff
@@ -5310,7 +5341,7 @@ int OSDMap::calc_desired_primary_distribution_osdsize_opt(
     }
     int write_ratio = 100 - read_ratio;
     ldout(cct, 30) << __func__ << " Pool: " << pid << " read ratio: " << read_ratio << " write ratio: " << write_ratio << dendl;
-    int num_osds = osds.size();
+    auto num_osds = osds.size();
```
Comment (cbodley) [inline src/osd/OSDMap.cc:5344]: this change from `int` to `auto` caused several signed comparison warnings:
```
[27/597] Building CXX object src/CMakeFiles/common-objs.dir/osd/OSDMap.cc.o
/home/cbodley/ceph/src/osd/OSDMap.cc: In member function ‘int OSDMap::calc_desired_primary_distribution_osdsize_opt(ceph::common::CephContext*, int64_t, const std::vector<long unsigned int>&, std::map<long unsigned int, float>&) const’:
/home/cbodley/ceph/src/osd/OSDMap.cc:5345:16: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
 5345 |     if (pg_num != (pool->get_pgp_num_mask() + 1)) {
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/cbodley/ceph/src/osd/OSDMap.cc:5373:26: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 5373 |       for (int i = 0 ; i < num_osds ; i++) {
      |                        ~~^~~~~~~~~~
/home/cbodley/ceph/src/osd/OSDMap.cc:5400:29: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 5400 |           if (osd_set_count == num_osds) {
      |               ~~~~~~~~~~~~~~^~~~~~~~~~~
/home/cbodley/ceph/src/osd/OSDMap.cc:5422:26: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 5422 |       for (int i = 0 ; i < num_osds; i++) {
      |                        ~~^~~~~~~~~~
/home/cbodley/ceph/src/osd/OSDMap.cc:5484:26: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 5484 |       for (int i = 0 ; i < num_osds ; i++) {
      |                        ~~^~~~~~~~~~
/home/cbodley/ceph/src/osd/OSDMap.cc: In member function ‘int OSDMap::calc_rbs_size_optimal(ceph::common::CephContext*, OSDMap&, int64_t, const pg_pool_t*, read_balance_info_t&) const’:
/home/cbodley/ceph/src/osd/OSDMap.cc:6589:23: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 6589 |   for (auto i = 0 ; i < num_osds ; i++) {
      |                     ~~^~~~~~~~~~
```

Comment (JoshSalomon) [reply src/osd/OSDMap.cc:5344]: I opened a PR to fix it, but I don't see any warnings when I build the project. What environment do you use for the build?

Comment (ljflores) [reply src/osd/OSDMap.cc:5344]: @cbodley which distro was this in ? Josh as raised a fix in a followup PR, but we want to make sure there are still no warnings.

Comment (cbodley) [reply src/osd/OSDMap.cc:5344]: sorry, this is fedora 39 with gcc 13.2.1

**Pull request merged**